<a href="https://colab.research.google.com/github/robitussin/CCRNFLRL_EXERCISES/blob/main/Exercise2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2

In [210]:
from math import sqrt, log
import pandas as pd

## Step 1: Enter your data

Add a new row after each pull: Pull number, Machine played ("A", "B", "C"), and Reward earned.

In [211]:
# data = pd.DataFrame({
#     "Pull": [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],          # Pull number
#     "Machine": ["A", "B", "C", "A", "B", "C", "A", "A", "A", "A", "A", "A", "A", "A",
#     "A", "B", "C", "A", "A", "A"], # Which machine was played
#     "Reward": [0, 0, 0, 0, 0, 0, 5, 2, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 5, 0]         # Reward received from that machine
# })

In [212]:
import pandas as pd

# Read Excel file
df_excel = pd.read_excel("slot_machine_tracker.xlsx")  # Replace with your actual file path

# Create the desired DataFrame structure
data = pd.DataFrame({
    "Pull": df_excel.iloc[:, 0],    # First column
    "Machine": df_excel.iloc[:, 1], # Second column
    "Reward": df_excel.iloc[:, 2]   # Third column
})

print(data)

    Pull Machine  Reward
0      1       A       0
1      2       B       0
2      3       C       0
3      4       A       0
4      5       B       0
5      6       C       0
6      7       A       5
7      8       A       2
8      9       A       0
9     10       A       0
10    11       A       0
11    12       A       0
12    13       A       1
13    14       A       1
14    15       A       1
15    16       B       0
16    17       C       0
17    18       A       0
18    19       A       5
19    20       A       0
20    21       A       0
21    22       A      10
22    23       A       0
23    24       A       0
24    25       A       2
25    26       A       0
26    27       A       0
27    28       A       0
28    29       A       0
29    30       A       1
30    31       A       1
31    32       A       0
32    33       A       0
33    34       A       0
34    35       A       0
35    36       B       0
36    37       C       5
37    38       C       5
38    39       C       0


# Step 2: Calculate UCB

In [213]:
def calculate_ucb(df):
    results = []
    total_pulls = len(df)
    machines = sorted(df["Machine"].unique())

    for machine in machines:
        machine_data = df[df["Machine"] == machine]
        n_j = len(machine_data)
        avg_reward = machine_data["Reward"].mean()
        exploration_bonus = sqrt((2 * log(total_pulls)) / n_j) if n_j > 0 else float("inf")
        ucb_score = avg_reward + exploration_bonus
        results.append({
            "Machine": machine,
            "Pulls": n_j,
            "Avg Reward": round(avg_reward, 2),
            "Exploration Bonus": round(exploration_bonus, 4),
            "UCB Score": round(ucb_score, 4)
        })
    return pd.DataFrame(results).sort_values(by="UCB Score", ascending=False)


# Step 3: Calculate and display


In [214]:
ucb_table = calculate_ucb(data)
print("📊 UCB Results Table")
print(ucb_table)

📊 UCB Results Table
  Machine  Pulls  Avg Reward  Exploration Bonus  UCB Score
0       A     30        0.97             0.5107     1.4774
2       C     16        0.75             0.6993     1.4493
1       B      4        0.00             1.3986     1.3986


# Step 4: Next machine suggestion

In [215]:
best_machine = ucb_table.iloc[0]["Machine"]
print(f"\n🎯 Recommended next machine to play: {best_machine}")


🎯 Recommended next machine to play: A


### Conclusion

After pulling to 3 different machine 50 times, machine A and C has a UCB Score of 1.4774 and 1.4493 respectively and machine B has a UCB Score of 1.3986. This means machine A and C are more likely has a chance of gaining a reward than machine B. Additionally, the average reward of 3 different machine are: machine A has 0.97, machine C has 0.75, and machine B has 0. This means that machine A has the biggest reward to be gained, machine B being the second, and machine C has more likely no rewards to be gained.